src/thinblock.cpp

```diff
@@ -22,20 +22,101 @@
 #include "txmempool.h"
 #include "util.h"
 #include "utiltime.h"
+#include "weakblock.h"
 
 using namespace std;
 
 static bool ReconstructBlock(CNode *pfrom, const bool fXVal, int &missingCount, int &unnecessaryCount);
 
+/*! Returns  the index into the weakblocks array if the block given as 'block' is a superset of that weak block. Returns a value
+  smaller than zero otherwise (no matching weak block found). */
+static int partOfLastWeak(const CBlock &block) {
+    LOCK(cs_weakblocks);
+    LogPrint("weakblocks", "Check whether block %s is delta on top of last or 2nd last weak block.\n", block.GetHash().GetHex());
+
+    if (!weakblocks.size()) {
+        LogPrint("weakblocks", "Currently no weak blocks -> Nope.\n");
+        return -1;
+    }
+
+    if (block.vtx.size() < 2) {
+        LogPrint("weakblocks", "Coinbase-transaction-only block -> Nope.\n");
+        return -1;
+    }
+
+    int result = weakblocks.size()-1;
+    CBlock *underlying = weakblocks[result];
+
+    if (underlying->GetHash() == block.GetHash()) {
+        LogPrint("weakblocks", "Block is identical to the last weak block -> comparing with next 2nd-last weak block.\n");
```

Comment:
Why continue looking further back in the chain of weak blocks if the block in question is identical to the last weak block? Seems like we should return index 0?

Reply:
Instead of returning a magic -1, returning a constant like NO_MATCHING_WBLOCK_FOUND would enhance the readability.

Reply:
@bderose: The point is that the last block will be the weakblock just received from another peer. The same data structure "weakblocks" is used to figure out when to send a deltathin block as well as to keep track of received blocks. But I take this as a doc bug. This needs to be explained better.

@MartensCedirc: Yes, fair point.